File: internal/radio/glyphs.go
package radio

import (
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/rivo/tview"
)

type glyphSet struct {
	play, stop, fail, song, notes, back, dot, shuffle, star, check, swatch string
	left, right, up, down                                                  string
	ellipsis                                                               string

	spinner []string

	gaugeFull, gaugeHalf, gaugeEmpty string

	vuLit, vuUnlit []string

	bands []string // spectrum bars, from quiet to full
}

var unicodeGlyphs = glyphSet{
	play: "▶", stop: "■", fail: "!", song: "♪", notes: "♫", back: "‹", dot: "·", shuffle: "🔀", star: "★", check: "✓", swatch: "██",
	left: "←", right: "→", up: "↑", down: "↓",
	ellipsis:  "…",
	spinner:   []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"},
	gaugeFull: "━", gaugeHalf: "╸", gaugeEmpty: "─",
	vuLit:   strings.Split("▁▂▂▃▃▄▅▅▆▆▇█", ""),
	vuUnlit: strings.Split("▁▂▂▃▃▄▅▅▆▆▇█", ""),
	bands:   strings.Split("▁▂▃▄▅▆▇█", ""),
}

var asciiGlyphs = glyphSet{
	play: ">", stop: "#", fail: "!", song: "~", notes: "*", back: "<", dot: "-", shuffle: "", star: "*", check: "*", swatch: "##",
	left: "<-", right: "->", up: "^", down: "v",
	ellipsis:  "~",
	spinner:   []string{"|", "/", "-", "\\"},
	gaugeFull: "=", gaugeHalf: "", gaugeEmpty: "-",
	vuLit:   strings.Split("||||||||||||", ""),
	vuUnlit: strings.Split("............", ""),
	bands:   strings.Split(".:|#", ""),
}

var glyphs = unicodeGlyphs

func applyGlyphs(ascii bool) {
	if ascii {
		glyphs = asciiGlyphs
	} else {
		glyphs = unicodeGlyphs
	}

	b := &tview.Borders
	if ascii {
		b.Horizontal, b.Vertical = '-', '|'
		b.TopLeft, b.TopRight, b.BottomLeft, b.BottomRight = '+', '+', '+', '+'
	} else {
		b.Horizontal, b.Vertical = '─', '│'
		b.TopLeft, b.TopRight, b.BottomLeft, b.BottomRight = '╭', '╮', '╰', '╯'
	}
	// Focus is shown with colour, not with tview's double lines.
	b.HorizontalFocus, b.VerticalFocus = b.Horizontal, b.Vertical
	b.TopLeftFocus, b.TopRightFocus = b.TopLeft, b.TopRight
	b.BottomLeftFocus, b.BottomRightFocus = b.BottomLeft, b.BottomRight
}

// detectASCII follows tcell's charset rules: where tcell falls back to ASCII,
// Unicode symbols would print as question marks.
func detectASCII() bool {
	if runtime.GOOS == "windows" {
		return false
	}
	if os.Getenv("TERM") == "linux" {
		return true
	}
	locale := os.Getenv("LC_ALL")
	if locale == "" {
		locale = os.Getenv("LC_CTYPE")
	}
	if locale == "" {
		locale = os.Getenv("LANG")
	}
	return !localeIsUTF8(locale)
}

func localeIsUTF8(locale string) bool {
	if locale == "C" || locale == "POSIX" {
		return false
	}
	if i := strings.IndexByte(locale, '@'); i >= 0 {
		locale = locale[:i]
	}
	i := strings.IndexByte(locale, '.')
	if i < 0 {
		return true // No charset given, e.g. "en_US" or empty: tcell assumes UTF-8.
	}
	charset := strings.ToLower(locale[i+1:])
	return charset == "utf-8" || charset == "utf8"
}

func spinnerFrame(t time.Time) string {
	return glyphs.spinner[int(t.UnixMilli()/100)%len(glyphs.spinner)]
}

File: rust/src/ui/glyphs.rs
use std::time::SystemTime;

pub struct Glyphs {
    pub ascii: bool,
    pub play: &'static str,
    pub stop: &'static str,
    pub fail: &'static str,
    pub song: &'static str,
    pub notes: &'static str,
    pub back: &'static str,
    pub dot: &'static str,
    pub shuffle: &'static str,
    pub star: &'static str,
    pub check: &'static str,
    pub swatch: &'static str,
    pub sleep: &'static str,
    /// The cursor bar's left edge.
    pub edge: &'static str,
    /// The card's status light.
    pub live: &'static str,
    pub left: &'static str,
    pub right: &'static str,
    pub up: &'static str,
    pub down: &'static str,
    pub ellipsis: &'static str,
    pub spinner: &'static [&'static str],
    pub gauge_full: &'static str,
    pub gauge_half: &'static str,
    pub gauge_empty: &'static str,
    /// Spectrum bars, from quiet to full.
    pub bands: &'static [&'static str],
    /// Box corners and lines: top-left, top-right, bottom-left, bottom-right,
    /// horizontal, vertical.
    pub border: [&'static str; 6],
}

pub const UNICODE: Glyphs = Glyphs {
    ascii: false,
    play: "▶",
    stop: "■",
    fail: "!",
    song: "♪",
    notes: "♫",
    back: "‹",
    dot: "·",
    shuffle: "🔀",
    star: "★",
    check: "✓",
    swatch: "██",
    sleep: "☾",
    edge: "▎",
    live: "•",
    left: "←",
    right: "→",
    up: "↑",
    down: "↓",
    ellipsis: "…",
    spinner: &["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"],
    gauge_full: "━",
    gauge_half: "╸",
    gauge_empty: "─",
    bands: &["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"],
    border: ["╭", "╮", "╰", "╯", "─", "│"],
};

pub const ASCII: Glyphs = Glyphs {
    ascii: true,
    play: ">",
    stop: "#",
    fail: "!",
    song: "~",
    notes: "*",
    back: "<",
    dot: "-",
    shuffle: "",
    star: "*",
    check: "*",
    swatch: "##",
    sleep: "z",
    edge: "",
    live: "o",
    left: "<-",
    right: "->",
    up: "^",
    down: "v",
    ellipsis: "~",
    spinner: &["|", "/", "-", "\\"],
    gauge_full: "=",
    gauge_half: "",
    gauge_empty: "-",
    bands: &[".", ":", "|", "#"],
    border: ["+", "+", "+", "+", "-", "|"],
};

impl Glyphs {
    pub fn spinner_frame(&self, t: SystemTime) -> &'static str {
        let ms = t.duration_since(SystemTime::UNIX_EPOCH).map_or(0, |d| d.as_millis());
        self.spinner[(ms / 100) as usize % self.spinner.len()]
    }
}

/// Follows tcell's charset rules, which the Go version used: where it fell
/// back to ASCII, Unicode symbols would print as question marks.
pub fn detect_ascii() -> bool {
    if cfg!(windows) {
        return false;
    }
    if std::env::var("TERM").is_ok_and(|t| t == "linux") {
        return true;
    }
    let locale = ["LC_ALL", "LC_CTYPE", "LANG"]
        .iter()
        .filter_map(|k| std::env::var(k).ok())
        .find(|v| !v.is_empty())
        .unwrap_or_default();
    !locale_is_utf8(&locale)
}

fn locale_is_utf8(locale: &str) -> bool {
    if locale == "C" || locale == "POSIX" {
        return false;
    }
    let locale = locale.split('@').next().unwrap_or("");
    match locale.split_once('.') {
        // No charset given, e.g. "en_US" or empty: UTF-8 is assumed.
        None => true,
        Some((_, charset)) => matches!(charset.to_ascii_lowercase().as_str(), "utf-8" | "utf8"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn locales() {
        assert!(locale_is_utf8("en_US.UTF-8"));
        assert!(locale_is_utf8("de_DE.utf8@euro"));
        assert!(locale_is_utf8(""));
        assert!(!locale_is_utf8("C"));
        assert!(!locale_is_utf8("en_US.ISO-8859-1"));
    }
}
